src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs

```diff
@@ -0,0 +1,85 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Graph
+{
+    public static class IDecryptableContentExtensions
+    {
+        private static readonly Lazy<int> AESInitializationVectorSize = new Lazy<int>(() =>
+        {
+            using (AesCryptoServiceProvider provider = new AesCryptoServiceProvider())
+            {
+                return provider.LegalBlockSizes[0].MinSize;
+            }
+        });
+
+        /// <summary>
+        /// Validates the signature and decrypted content attached with the notification.
+        /// </summary>
+        /// <typeparam name="T">Type to deserialize the data to.</typeparam>
+        /// <param name="certificateProvider">Certificate provider to decrypt the content. The first parameter is the certificate ID provided when creating the subscription. The second is the certificate thumbprint. The certificate WILL be disposed at the end of decryption.</param>
+        /// <returns>Decrypted content as the provided type.</returns>
```

Comment:
it feels that this generic type should have more constraints. Basically the type should be one of the types available in the SDK or a collection of a type available in the SDK. Let me know what you think

Reply:
One challenge here I see is how will the dev know what type is the type parameter. If the listener URL only receives notifications for a specific type, it will work fine. If the listener is used for subscriptions that target different types, they won't be able specify the type.  We could:
* leave it to dev to maintain a mapping of subscriptionId to the type it returns so they can use that type parameter for `Decrypt`. OR
* We add Entity (base for all Graph entities) partial to core, and provide something like we have in [DerivedTypeConverter](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/dev/src/Microsoft.Graph.Core/Serialization/DerivedTypeConverter.cs) to deserialize to the type described in odata.type. It is deserialized into a descendant of Entity, and should have odata.type set on it.

Reply:
or maybe we keep a very simple constraint like where T : class, new ?

Reply:
class makes sense, but not new. If a content data entity is a heterogenous collection with a common base type, we'd want to be able to specify the base type.